### ISSUE TITLE
LB-IPAM: Unable to set single addresses as a pool

### ISSUE THREAD

**Issue by samip5** [cilium/cilium#24351]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I tried to create a pool with only single IPv4 (/32) and IPv6 (/128) addresses inside it.
I expected it to see that the pool has two IP addresses in total, but it doesn't seem to get that.

Instead it's having very hard time allocating IPs from it.

### Cilium Version

cilium-cli: v0.12.1 compiled with go1.18.5 on linux/amd64
cilium image (default): v1.12.0
cilium image (stable): v1.13.0
cilium image (running): v1.13.0

### Kernel Version

5.15.0-1025-raspi

### Kubernetes Version

```
{
  "clientVersion": {
    "major": "1",
    "minor": "26",
    "gitVersion": "v1.26.1",
    "gitCommit": "8f94681cd294aa8cfd3407b8191f6c70214973a4",
    "gitTreeState": "archive",
    "buildDate": "2023-02-06T14:43:47Z",
    "goVersion": "go1.19.4",
    "compiler": "gc",
    "platform": "linux/amd64"
  },
  "kustomizeVersion": "v4.5.7",
  "serverVersion": {
    "major": "1",
    "minor": "25",
    "gitVersion": "v1.25.6+k3s1",
    "gitCommit": "9176e03c5788e467420376d10a1da2b6de6ff31f",
    "gitTreeState": "clean",
    "buildDate": "2023-01-26T00:30:33Z",
    "goVersion": "go1.19.5",
    "compiler": "gc",
    "platform": "linux/arm64"
  }
}
```

### Sysdump

Too large; It's [here](https://s3.skym.fi/public/cilium-sysdump-20230313-191606.zip)

### Relevant log output

```shell
level=info msg="Notify ingress controller for service ingress" subsys=ingress-controller
level=error msg="Failed to process Ingress event, skipping: {oldIngress:0x4000171f00 newIngress:0x40015bc100}" subsys=ingress-controller
level=error msg="Unable to allocate IP" error="provided IP is not in the valid range. The range of valid IPs is 185.218.193.58/32" subsys=lbipam
level=error msg="Unable to allocate IP" error="provided IP is not in the valid range. The range of valid IPs is 2001:67c:1104:e03e::443/128" subsys=lbipam
level=info msg="Notify ingress controller for service ingress" subsys=ingress-controller
level=error msg="Unable to allocate IP" error="provided IP is not in the valid range. The range of valid IPs is 185.218.193.58/32" subsys=lbipam
level=error msg="Unable to allocate IP" error="provided IP is not in the valid range. The range of valid IPs is 2001:67c:1104:e03e::443/128" subsys=lbipam
```

### Anything else?

This is valid way for MetalLB, and I think it should work with Cilium too.

The pool manifest can be found [here](https://github.com/samip5/k3s-test-cluster/blob/6fa03734ed98edbaf4a86d79c7aa0bec3d70fe06/kubernetes/apps/cilium-system/cilium/cilium-lb-pool.yaml)

The cilium ingress svc looks like this:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    io.cilium/lb-ipam-ips: 185.218.193.58,2001:67c:1104:e03e::443
    meta.helm.sh/release-name: cilium
    meta.helm.sh/release-namespace: kube-system
  creationTimestamp: "2023-03-13T16:52:43Z"
  labels:
    app.kubernetes.io/managed-by: Helm
    cilium.io/ingress: "true"
    helm.toolkit.fluxcd.io/name: cilium
    helm.toolkit.fluxcd.io/namespace: kube-system
  name: cilium-ingress
  namespace: kube-system
  resourceVersion: "2315905"
  uid: 2b9c0a9c-b58a-45dd-bd47-1e00f21533cc
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.46.159.107
  clusterIPs:
  - 10.46.159.107
  - 2001:67c:1104:e03e::c94b
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: RequireDualStack
  ports:
  - name: http
    nodePort: 31019
    port: 80
    protocol: TCP
    targetPort: 80
  - name: https
    nodePort: 31816
    port: 443
    protocol: TCP
    targetPort: 443
  sessionAffinity: None
  type: LoadBalancer
status:
  conditions:
  - lastTransitionTime: "2023-03-13T16:52:43Z"
    message: ""
    reason: satisfied
    status: "True"
    type: io.cilium/lb-ipam-request-satisfied
  loadBalancer: {}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by borkmann** [comment]
Cc @dylandreimerink could you take a look? thx

**Comment by dylandreimerink** [comment]
The reason why this doesn't work comes down to an implementation detail. LB-IPAM uses an existing package for the IP allocation which reserves the first and last IP since these typically can't be used. https://github.com/cilium/ipam/blob/master/service/ipallocator/allocator.go#L56

But this also applies when making the range of /32 or /31, which does not make sense. We need to do some work on the IP allocation logic to fix this one.

For now, a temporary workaround would be to add a /30 to your pool, this does make 2 IPs available not 1 however, so not exactly what you are after.

**Comment by samip5** [comment]
>But this also applies when making the range of /32 or /31, which does not make sense. 

Which portion doesn't make sense? The fact that it wants to eat two IPs even when the pool is /32 or /31?

> For now, a temporary workaround would be to add a /30 to your pool, this does make 2 IPs available not 1 however, so not exactly what you are after.

Unfortunately, public IPv4's are not free, so no I don't have a /30 to add to the pool.

**Comment by hainamtvvt** [comment]
@samip5 you can allocate IP in file service type LB. Example 
```
apiVersion: v1
kind: Service
metadata:
  name: service-yellow
  namespace: tenant-c
  labels:
    color: yellow
spec:
  type: LoadBalancer
  LoadBalancerIP: 20.0.10.101
  ports:
  - port: 1234
 ```

**Comment by samip5** [comment]
> @samip5 you can allocate IP in file service type LB. Example 
> ```
> apiVersion: v1
> kind: Service
> metadata:
>   name: service-yellow
>   namespace: tenant-c
>   labels:
>     color: yellow
> spec:
>   type: LoadBalancer
>   LoadBalancerIP: 20.0.10.101
>   ports:
>   - port: 1234
>  ```

Unfortunately, that doesn't work when Cilium doesn't understand the pool properly, with /32 prefix which is a single IPv4 and instead says there's no IPv4 addresses.

**Comment by 00Asgaroth00** [comment]
+1

I'm currently testing/learning with K3S v1.26.2+k3s1 and cilium version 1.13.0/1.13.1

I'm having a similar issue:

Applying:
```
apiVersion: "cilium.io/v2alpha1"
kind: CiliumLoadBalancerIPPool
metadata:
    name: "global-pool"
spec:
    cidrs:
        - cidr: "10.6.21.220/32"
        - cidr: "10.6.21.221/32"
```
results in:
```
~ $ sudo -E kubectl get ippools
NAME          DISABLED   CONFLICTING   IPS AVAILABLE   AGE
global-pool   false      False         0               58s
```
A describe shows:
```
~ $ sudo -E kubectl describe ippools/global-pool
Name:         global-pool
Namespace:
Labels:       <none>
Annotations:  <none>
API Version:  cilium.io/v2alpha1
Kind:         CiliumLoadBalancerIPPool
Metadata:
  Creation Timestamp:  2023-03-20T14:27:27Z
  Generation:          1
  Managed Fields:
    API Version:  cilium.io/v2alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
          .:
          k:{"type":"io.cilium/conflict"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
          k:{"type":"io.cilium/ips-available"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
          k:{"type":"io.cilium/ips-total"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
          k:{"type":"io.cilium/ips-used"}:
            .:
            f:lastTransitionTime:
            f:message:
            f:observedGeneration:
            f:reason:
            f:status:
            f:type:
    Manager:      cilium-operator-lb-ipam
    Operation:    Update
    Subresource:  status
    Time:         2023-03-20T14:27:27Z
    API Version:  cilium.io/v2alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:cidrs:
        f:disabled:
    Manager:         kubectl-client-side-apply
    Operation:       Update
    Time:            2023-03-20T14:27:27Z
  Resource Version:  4858
  UID:               a73ebe54-51e8-4106-8cb1-0aa6e5ae88c6
Spec:
  Cidrs:
    Cidr:    10.6.21.220/32
    Cidr:    10.6.21.221/32
  Disabled:  false
Status:
  Conditions:
    Last Transition Time:  2023-03-20T14:27:27Z
    Message:
    Observed Generation:   1
    Reason:                resolved
    Status:                False
    Type:                  io.cilium/conflict
    Last Transition Time:  2023-03-20T14:27:27Z
    Message:               0
    Observed Generation:   1
    Reason:                noreason
    Status:                Unknown
    Type:                  io.cilium/ips-total
    Last Transition Time:  2023-03-20T14:27:27Z
    Message:               0
    Observed Generation:   1
    Reason:                noreason
    Status:                Unknown
    Type:                  io.cilium/ips-available
    Last Transition Time:  2023-03-20T14:27:27Z
    Message:               0
    Observed Generation:   1
    Reason:                noreason
    Status:                Unknown
    Type:                  io.cilium/ips-used
Events:                    <none>
```

**Comment by DJAlPee** [comment]
Stumpled on this, too...
Would be great to have something similar to MetalLB: https://metallb.universe.tf/apis/#metallb.io/v1beta1.IPAddressPool

Maybe, this could be used for "inspiration" 😉

**Comment by omniproc** [comment]
Running into the same issue here. More range-logic inspiration from Kube-VIP: https://github.com/kube-vip/kube-vip-cloud-provider#ip-address-functionality

**Comment by j-gooding** [comment]
Having the same issue, unable to allocate a cidr of /32. With the introduction of L2 announcements (https://docs.cilium.io/en/latest/network/l2-announcements/) in v1.14 this is going to be an even bigger issue. If the goal is to help support smaller on-perm deployments such as campus networks, then it is critical to be able to allocate single IP cidrs or arbitrary ranges as described here: https://github.com/cilium/cilium/issues/24351#issuecomment-1479660360. Public IP's are very valuable and not handed out on a whim, they're often allocated one at a time or in arbitrary ranges. 

> L2 Announcements is a feature which makes services visible and reachable on the local area network. This feature is primarily intended for on-premises deployments within networks without BGP based routing such as office or campus networks.

I would go so far as to say the L2 announcements feature is not very useful in real world practice without an improved LB-IPAM capability. To me it would be important to have some kind of fix or clearly documented safe workaround for this to be released with the L2 feature.

**Comment by w7team** [comment]
希望可以支持单ip使用

**Comment by MTRNord** [comment]
Hitting the same. This is mostly annoying for smaller deployments I guess but it is annoying.

**Comment by dorkamotorka** [comment]
+1

**Comment by alessiodionisi** [comment]
Same issues here

**Comment by CallMeFoxie** [comment]
With multi-pool on the rise we are fine announcing /32 and /128 on our network (BGP), this vastly limits our possible usage as it wastes a lot of IPs... any update? Our usecase is that we have separate pool for specific uses (NAT, fixed public IPs for some pods) :/

**Comment by dylandreimerink** [comment]
> With multi-pool on the rise we are fine announcing /32 and /128 on our network (BGP), this vastly limits our possible usage as it wastes a lot of IPs... any update? 

PR #26488 is going to fix this, all review feedback has been processed, I am keeping it up-to-date (rebases on main and CI). It is currently waiting on re-review from two people who are away for a bit. If all goes well it will get merged soon™. 

I have not forgotten about this!

**Comment by CallMeFoxie** [comment]
> > With multi-pool on the rise we are fine announcing /32 and /128 on our network (BGP), this vastly limits our possible usage as it wastes a lot of IPs... any update?
> 
> PR #26488 is going to fix this, all review feedback has been processed, I am keeping it up-to-date (rebases on main and CI). It is currently waiting on re-review from two people who are away for a bit. If all goes well it will get merged soon™.
> 
> I have not forgotten about this!

Is this just for lb-ipam though? We use it for different workloads on actual pods.

**Comment by dylandreimerink** [comment]
> Is this just for lb-ipam though?

Yes, this is a LB-IPAM specific issue, I recommend opening a separate issue for multi-pool IPAM since I believe this currently isn't being tracked. (Looking at #25470).

That being said, the work in the above PR has been made generic enough that is should be usable to resolve the same issue in multi-pool if we decide the same allocator logic also works in that use-case. So with a bit of luck, addressing the issue can be faster than with LB-IPAM.
